lib/temporal/worker.rb

```diff
@@ -0,0 +1,72 @@
+require 'async'
+require 'temporal/bridge'
+require 'temporal/data_converter'
+require 'temporal/failure_converter'
+require 'temporal/payload_converter'
+require 'temporal/runtime'
+require 'temporal/worker/activity'
+require 'temporal/worker/thread_pool_executor'
+
+module Temporal
+  class Worker
+    # TODO: Add worker interceptors
+    def initialize(
+      connection,
+      namespace,
+      task_queue,
+      payload_converter: Temporal::PayloadConverter::DEFAULT,
+      payload_codecs: [],
+      failure_converter: Temporal::FailureConverter::DEFAULT,
+      activity_executor: nil,
+      max_concurrent_activities: 100
+    )
+      @running = false
+      @mutex = Mutex.new
+      @runtime = Temporal::Runtime.instance
+      activity_executor ||= ThreadPoolExecutor.new(max_concurrent_activities)
+      data_converter = DataConverter.new(
+        payload_converter: payload_converter,
+        payload_codecs: payload_codecs,
+        failure_converter: failure_converter,
+      )
+      core_worker = Temporal::Bridge::Worker.create(
+        @runtime.core_runtime,
+        connection.core_connection,
+        namespace,
+        task_queue,
+      )
+      @activity_worker = Worker::Activity.new(core_worker, data_converter, activity_executor)
```

Comment:
What are the Ruby norms on visibility/accessibility of these things? We want to hide things like activity workers from a user, but is this considered public now? Does that mean we're beholden to compatibility guarantees when changing it?

Reply:
Since everything is in a global namespace there's no good way of hiding classes from the public API. We can make nested classes (such as a `Temporal::Worker::Activity`) private, however this would mean that nothing apart from the `Temporal::Worker` would be able to access it. While it might work in this case, people tend to stay away from it as it means you can't really access the class in any other way.

From what I've seen these are normally kept accessible anyways and gem users are not expected to interact with these directly. We can probably mark these as private in our YARD docs to make it clear that there are zero compatibility guarantees for these classes.

Reply:
How does a user know that, say, `Temporal::Error::ApplicationError` is stable code they can rely on but `Temporal::Worker::Activity` is not?

Reply:
To be honest with you I don't think there's a clear distinction. I check source code of a few popular gems (including `rails`) and there isn't a single class (and there are many) I could find that is clearly private. It's probably down to docs to steer users away from what's clearly private (and therefore doesn't guarantee API stability)

Reply:
:+1: I only bring this up because we've been burned by being beholden to public API that should not have been public. We also try to offer users guarantees that we almost never change public API. Whatever way you can reconcile these concerns best for the Ruby community works for me.

But I can say from experience during the building and maintenance of this project, these private pieces will probably get refactored a lot.